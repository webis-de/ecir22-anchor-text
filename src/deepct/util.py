def clean_qtokens(qtokens):
    ret = {}
    blocklist = load_entrypage_queries()

    for doc, doc_tokens in qtokens.items():
        if len(doc_tokens.keys()) == 0 or len([i for i in doc_tokens.keys() if i in blocklist])/len(doc_tokens.keys()) > 0.2:
            continue
        ret[doc] = doc_tokens

    print('From qtokens with ' + str(len(qtokens)) + ' documents, I filtered terms that may cause train/test overlap so that I now have ' + str(len(ret)) + ' documents.')

    return ret


def load_entrypage_queries():
    from glob import glob
    import importlib
    text_clean = importlib.import_module('train-deep-ct').text_clean

    ret = []
    for f in glob('../../Data/navigational-topics-and-qrels-ms-marco-v1/topics*.tsv'):
        with open(f, 'r') as f:

            for l in f:
                l = l.split('\t')
                if len(l) != 2:
                    raise ValueError('Can not handle ' + str(l))

                ret += text_clean(l[1].strip(), True, True).split(' ')

    return set(ret)


def sample_anchors(anchors_per_doc):
    import json
    import importlib
    text_clean = importlib.import_module('train-deep-ct').text_clean
    blocklist = load_entrypage_queries()
    ret = set()
    for anchors in anchors_per_doc:
        for anchor in set(json.loads(anchors)['anchorText']):
            anchor_terms = text_clean(anchor.strip(), True, True).split(' ')
            if len(anchor_terms) > 0 and len([i for i in anchor_terms if i in blocklist])/len(anchor_terms) < 0.51:
                ret |= set([anchor])

    return ret
