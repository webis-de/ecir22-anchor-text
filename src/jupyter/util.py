import spacy
from copy import deepcopy
import random
import json
import numpy as np
from nltk.stem.porter import PorterStemmer
stemmer = PorterStemmer()
import itertools
nlp = None
ALL_PERMUTATIONS=set([i for i in itertools.permutations([0,0,0,0,0,1,1,1,1,1])])


def distinct_terms(processed_terms):
    return set([i['stemmed'] for i in processed_terms])


def extract_terms(sentence):
    global nlp
    if nlp is None:
        nlp = spacy.load("en_core_web_sm", exclude=["parser", "tagger", "ner", "attribute_ruler", "lemmatizer", "tok2vec"])
    ret = []
    for i in nlp(sentence):
        if not i.is_punct:
            ret += [{'term': str(i), 'stemmed': stemmer.stem(str(i))}]
    return ret


def json_lines(filename):
    from tqdm import tqdm
    import json
    with open(filename, 'r') as f:
        for i in tqdm(f):
            yield json.loads(i)


def enrich_presentation(presentation):
    ret = deepcopy(presentation)
    if 'contents' in ret:
        ret['contentsProcessed'] = extract_terms(ret['contents'])
    if 'anchorText' in ret:
        ret['anchorTextProcessed'] = []
        for anchor in ret['anchorText']:
            ret['anchorTextProcessed'] += [extract_terms(anchor)]

    return ret


def __terms_for_counting(line):
    if 'contentsProcessed' in line:
        yield line['contentsProcessed']
    if 'anchorTextProcessed' in line:
        for anchor in line['anchorTextProcessed']:
            yield anchor


def __flatten(terms):
    ret = {}
    for term in terms.keys():
        term_with_count = [(i,c) for i,c in terms[term].items()]

        most_term = sorted(term_with_count, key=lambda i: i[1], reverse=True)[0][0]
        ret[most_term] = sum(i[1] for i in term_with_count)

    return ret


def count_terms(lines):
    ret = {}
    for line in lines:
        if type(line) == str:
            line = json.loads(line)
        for terms in __terms_for_counting(line):
            for term in terms:
                if term['stemmed'] not in ret:
                    ret[term['stemmed']] = {}
                if term['term'] not in ret[term['stemmed']]:
                    ret[term['stemmed']][term['term']] = 0

                ret[term['stemmed']][term['term']] += 1

    return __flatten(ret)
            
def distribution_distinct_terms(lines):
    ret = {}

    for line in lines:
        if type(line) == str:
            line = json.loads(line)

        terms_in_line = str(len(distinct_terms(line['contentsProcessed'])))
        
        if int(terms_in_line) <= 0:
            continue

        if terms_in_line not in ret:
            ret[terms_in_line] = 0

        to_add = 1
        if 'target_document' in line and type(line['target_document']) == list:
            to_add = len(line['target_document'])
        ret[terms_in_line] += to_add

    return ret

def anchor_text_distribution_distinct_terms(lines, limit=None):
    ret = []
    for line in lines:
        if type(line) == str:
            line = json.loads(line)

        if limit != None and limit< len(line['anchorTextProcessed']):
            continue

        for anchor in line['anchorTextProcessed']:
            ret += [{'contentsProcessed': anchor}]

    return distribution_distinct_terms(ret)

def aggregated_anchor_text_distribution_distinct_terms(lines, limit=None):
    ret = []
    for line in lines:
        if type(line) == str:
            line = json.loads(line)
        contents = []
        if limit != None and limit< len(line['anchorTextProcessed']):
            continue

        for anchor in line['anchorTextProcessed']:
            contents += anchor
        ret += [{'contentsProcessed': contents}]

    return distribution_distinct_terms(ret)

def build_probability_arrays(first, second):
    first_ret = []
    second_ret = []
    first_sum = sum(first.values())
    second_sum = sum(second.values())

    max_id = max([int(i) for i in first.keys()] + [int(i) for i in second.keys()])
    for i in range(1, max_id +1):
        first_ret += [first.get(str(i), 0) / first_sum]
        second_ret += [second.get(str(i), 0) / second_sum]

    return first_ret, second_ret


def __combine_terms(docs):
    ret = {}
    for doc in docs:
        for term, count in doc.items():
            if term not in ret:
                ret[term] = 0
            ret[term] += count
    return ret


def build_probability_arrays_terms(first, second):
    first_ret = []
    second_ret = []
    first_sum = sum(first.values())
    second_sum = sum(second.values())

    terms = set([i for i in first.keys()] + [i for i in second.keys()])

    for i in sorted(terms):
        first_ret += [first.get(str(i), 0) / first_sum]
        second_ret += [second.get(str(i), 0) / second_sum]

    return first_ret, second_ret


def run_homogenity_experiments(json_line, num_splits=10):
    batches = combine_retrieved_documents_to_batches(json_line, num_splits)

    for batch in batches:
        if sum(batch.values()) == 0:
            return []

    topic = json.loads(json_line)['topic']
    ret = []

    for slice_ids in ALL_PERMUTATIONS:
        parts = [[], []]

        for i in range(num_splits):
            parts[slice_ids[i]] += [batches[i]]

        assert len(parts) == 2
        assert len(parts[0]) == 5
        assert len(parts[1]) == 5

        part_0 = __combine_terms(parts[0])
        part_1 = __combine_terms(parts[1])

        part_0, part_1 = build_probability_arrays_terms(part_0, part_1)

        ret += [{
            'topic': topic,
            'jensen_shannon_distance': jensen_shannon_distance(part_0, part_1),
        }]

    return ret


def evaluate_retrieval_homogenity(directory, index, retrieval_model, processes=20):
    from multiprocessing import Pool
    with Pool(processes=processes) as pool:
        lines = [i for i in open(directory + '/run.' + index + '.' + retrieval_model + '.txt.jsonl')]
        tmp_ret = [i for i in pool.map(run_homogenity_experiments, lines, 1)]
        ret = []
        for li in tmp_ret:
            for i in li:
                i['index'] = index
                i['retrieval_model'] = retrieval_model
                ret += [i]

        return ret


def combine_retrieved_documents_to_batches(json_line, num_splits=10):
    json_line = json.loads(json_line)
    documents = [i['documentRepresentation'] for i in json_line['documents'] if i is not None and 'documentRepresentation' in i and
 i['documentRepresentation'] is not None and len(i['documentRepresentation'].keys()) >= 2]
    random.Random(num_splits).shuffle(documents)
    
    return [__combine_terms(i) for i in np.array_split(documents, num_splits)]


def jensen_shannon_distance(p, q):
    from scipy.spatial.distance import jensenshannon
    return jensenshannon(p,q)

def kullback_leibler_divergence(p,q):
    from scipy.stats import entropy
    assert p != None
    assert q != None # scipy.stats.entropy calculate KL-divergence if p and q are both not none

    return entropy(p, q)

