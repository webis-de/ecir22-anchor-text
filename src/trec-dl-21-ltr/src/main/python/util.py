def write_run_file(model_file, feature_file, system_name, out_file):
    with open(out_file, 'w') as f:
        f.write(run_file_content(model_file, feature_file, system_name))

def run_file_content(model_file, feature_file, system_name):
    from lightgbm import Booster
    model = Booster(model_file=model_file)
    ret = ''

    to_rerank = load_ranking_data(feature_file)
    for topic_to_rerank in to_rerank:
        doc_to_score = []
        y_predictions = model.predict(topic_to_rerank['documentFeatures'])
        assert len(y_predictions) == len(topic_to_rerank['documents'])

        for i in range(len(topic_to_rerank['documents'])):
            doc_to_score += [{'documentId': topic_to_rerank['documents'][i], 'score': y_predictions[i]}]

        doc_to_score = sorted(doc_to_score, key=lambda i: i['score'], reverse=True)
        ret += topic_as_run_file(topic_to_rerank['topic'], doc_to_score, system_name)

    return ret.strip()

def topic_as_run_file(topic_nr, doc_to_score, system_name):
    ret = ''

    for i in range(len(doc_to_score)):
        ret += topic_nr + ' Q0 ' + doc_to_score[i]['documentId'] + ' ' + str(i+1) + ' ' + str(doc_to_score[i]['score']) + ' ' + system_name + '\n'

    return ret

def load_ranking_data(file_name):
    import json
    ret = []
    with open(file_name, 'r') as f:
        for l in f:
            l = json.loads(l)

            ret += [{
                'topic': l['topic'],
                'documents': [i['documentId'] for i in l['documents']],
                'documentFeatures': as_np_array(l['documents']),
            }]
    return ret


def as_np_array(docs):
    import numpy as np
    ret = []
    for doc in docs:
        last_feature=1
        features = doc['features']
        # relevance label and query id
        ret_features = [-100, -100]
        for feature_num in features:
            if (int(feature_num) -1) != last_feature:
                raise ValueError('Could not handle...')

            ret_features += [features[feature_num]]
            last_feature = int(feature_num)
        ret += [ret_features]

    return np.array(ret)

