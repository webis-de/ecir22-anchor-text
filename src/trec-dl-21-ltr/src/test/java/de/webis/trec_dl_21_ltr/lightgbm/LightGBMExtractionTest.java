package de.webis.trec_dl_21_ltr.lightgbm;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Test;
import org.mockito.Matchers;
import org.mockito.Mockito;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

import com.fasterxml.jackson.databind.ObjectMapper;

import io.anserini.eval.RelevanceJudgments;
import lombok.SneakyThrows;

public class LightGBMExtractionTest {

  private static final String DOCUMENT_FEATURES_FILE = "src/test/resources/feature-extraction/ms-marco-document-features.jsonl";
  private static final String QUERY_FEATURES_FILE = "src/test/resources/feature-extraction/ms-marco-query-features.jsonl";
  private static final String QUERY_DOCUMENT_FEATURES_DIR = "src/test/resources/feature-extraction/";

  @Test
  public void assert50DistinctFeaturesAreUsed() {
    Assert.assertEquals(50, new HashSet<>(App.FEATURES).size());
  }
  
  @Test
  public void assert121DistinctV1FeaturesAreUsed() {
    Assert.assertEquals(121, new HashSet<>(App.FEATURES_V1).size());
  }
  
  @Test
  public void assert112DistinctV1WithoutOrcasFeaturesAreUsed() {
    Assert.assertEquals(112, new HashSet<>(App.FEATURES_V1_NO_ORCAS).size());
  }
 
  @Test
  public void assert49DistinctV1WithoutAnchorFeaturesAreUsed() {
    Assert.assertEquals(49, new HashSet<>(App.FEATURES_V1_NO_ANCHOR).size());
  }
  
  @Test
  public void assert40DistinctV1WithoutAnchorWithoutOrcasFeaturesAreUsed() {
    Assert.assertEquals(40, new HashSet<>(App.FEATURES_V1_NO_ANCHOR_NO_ORCAS).size());
  }
  
  @Test
  @SneakyThrows
  public void approveFeatures() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES.size(); i++) {
      actual.put(i+2, App.FEATURES.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesWithoutAnchors() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_NO_ANCHOR.size(); i++) {
      actual.put(i+2, App.FEATURES_NO_ANCHOR.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesV1() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_V1.size(); i++) {
      actual.put(i+2, App.FEATURES_V1.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesV1NoOrcas() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_V1_NO_ORCAS.size(); i++) {
      actual.put(i+2, App.FEATURES_V1_NO_ORCAS.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesV1NoAnchorNoOrcas() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_V1_NO_ANCHOR_NO_ORCAS.size(); i++) {
      actual.put(i+2, App.FEATURES_V1_NO_ANCHOR_NO_ORCAS.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesV1NoAnchor() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_V1_NO_ANCHOR.size(); i++) {
      actual.put(i+2, App.FEATURES_V1_NO_ANCHOR.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }  

  @Test
  public void testRunFile() {
    Integer relevanceLabel = null;
    Map<String, Object> features = Map.of(
      "f1", 0.01,
      "f2", 0.02,
      "f3", 0.03
    );
    
    String expected = "0 1:-4 2:0.02 3:0.01";
    String queryId = "-4";
    String documentId = "doc-2";
    
    String actual = LightGBMExtraction.formatLine(queryId, documentId, features, Arrays.asList("f2","f1"), relevanceLabel);
    
    Assert.assertEquals(expected, actual);
  }
  
  @Test
  public void testRunFile2() {
    Integer relevanceLabel = null;
    Map<String, Object> features = Map.of(
      "f1", 0.01,
      "f2", 0.02,
      "f3", 0.03
    );
    
    String expected = "0 1:queryId 2:0.01 3:0.02 4:0.03";
    String queryId = "queryId";
    String documentId = "doc";
    
    String actual = LightGBMExtraction.formatLine(queryId, documentId, features, Arrays.asList("f1","f2", "f3"), relevanceLabel);
    
    Assert.assertEquals(expected, actual);
  }
  
  @Test(expected = RuntimeException.class)
  public void testExtractionOfNonExistingFeature() {
    Integer relevanceLabel = null;
    Map<String, Object> features = Map.of(
      "f1", 0.01,
      "f2", 0.02,
      "f3", 0.03
    );
    String queryId = "queryId";
    String documentId = "doc";
    
    LightGBMExtraction.formatLine(queryId, documentId, features, Arrays.asList("f1","f2", "f4"), relevanceLabel);
  }

  @Test
  @SneakyThrows
  public void approveTestFeatureFile() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile(DOCUMENT_FEATURES_FILE)
    	.queryFeaturesFile(QUERY_FEATURES_FILE)
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopics())
    	.validationTopics(Arrays.asList(dummyTopics()))
    	.testTopics(dummyTopics())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("2021_trec_dl_test")));
  }

  @Test
  @SneakyThrows
  public void approveTestRerankFile() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile(DOCUMENT_FEATURES_FILE)
    	.queryFeaturesFile(QUERY_FEATURES_FILE)
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopics())
    	.validationTopics(Arrays.asList(dummyTopics()))
    	.testTopics(dummyTopics())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("2021_trec_dl_test.rerank")));
  }
  
  @Test
  @SneakyThrows
  public void approveValidationFeatureFile() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile(DOCUMENT_FEATURES_FILE)
    	.queryFeaturesFile(QUERY_FEATURES_FILE)
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopics())
    	.validationTopics(Arrays.asList(dummyTopics()))
    	.testTopics(dummyTopics())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("docv2_dev1")));
  }

  @Test
  @SneakyThrows
  public void approveTrainFeatureFile() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile(DOCUMENT_FEATURES_FILE)
    	.queryFeaturesFile(QUERY_FEATURES_FILE)
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopics())
    	.validationTopics(Arrays.asList(dummyTopics()))
    	.testTopics(dummyTopics())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("docv2_train")));
  }
  
  @Test
  @SneakyThrows
  public void approveTrainFeatureFileMsMarcoV1() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile("src/test/resources/feature-extraction/ms-marco-v1-document-features.jsonl")
    	.queryFeaturesFile("src/test/resources/feature-extraction/ms-marco-v1-query-features.jsonl")
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES_V1)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopicsV1())
    	.validationTopics(Arrays.asList(dummyTopicsV1()))
    	.testTopics(dummyTopicsV1())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("docv2_train")));
  }
  
  private Map<Integer, Set<String>> dummyTopics() {
    return Map.of(
      1000006, Set.of("msmarco_doc_09_1282503587", "msmarco_doc_14_800591714"),
      1000412, Set.of("msmarco_doc_21_243726663", "msmarco_doc_21_243753677")
    );
  }
  
  private Map<Integer, Set<String>> dummyTopicsV1() {
    return Map.of(
      1090311, Set.of("D99747", "D99748")
    );
  }

  private RelevanceJudgments relevanceJudgments() {
    RelevanceJudgments relevanceJudgments = Mockito.mock(RelevanceJudgments.class);
    Mockito.when(relevanceJudgments.getRelevanceGrade(Matchers.any(), Matchers.any())).thenAnswer(new Answer<Integer>() {
      @Override
      public Integer answer(InvocationOnMock invocation) throws Throwable {
        return Arrays.asList("msmarco_doc_09_1282503587", "msmarco_doc_21_243753677", "D99747").contains(invocation.getArguments()[1]) ? 1 : 0;
      }
    });
	Mockito.when(relevanceJudgments.isDocJudged(Matchers.any(), Matchers.any())).thenReturn(true);
    
    return relevanceJudgments;
  }
}
