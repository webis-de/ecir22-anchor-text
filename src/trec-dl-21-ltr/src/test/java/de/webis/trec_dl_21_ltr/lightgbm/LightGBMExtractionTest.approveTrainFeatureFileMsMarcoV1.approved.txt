1 1:1090311 2:6.0166183 3:5.607375 4:4.092708 5:4.706899 6:7.244738 7:14.181548 8:5.717017 9:1.0725099 10:321.0 11:0.0 12:0.0 13:0.0 14:0.0 15:0.0 16:0.0 17:0.0 18:0.0 19:0.0 20:4.904618 21:10.616128 22:1.9948121 23:2.8757951 24:2.9731674 25:11.334622 26:4.8564687 27:6.554533 28:1.0 29:7.580994 30:14.336913 31:7.547449 32:9.949312 33:7.1680427 34:9.934192 35:7.566821 36:5.081198 37:452.0 38:8.7661495 39:14.894001 40:7.48988 41:9.916091 42:7.9484506 43:13.203759 44:8.531042 45:5.5594 46:520.0 47:9.212866 48:14.471718 49:7.7888584 50:10.025941 51:9.267184 52:14.313749 53:8.921597 54:5.686173 55:1012.0 56:8.54594 57:13.641667 58:7.454386 59:9.679138 60:8.836189 61:14.017209 62:8.2871 63:5.5655217 64:724.0 65:9.348539 66:14.053547 67:7.7752004 68:9.936002 69:8.987881 70:14.057504 71:9.05758 72:5.7526174 73:949.0 74:8.826133 75:13.772553 76:7.3755856 77:9.638531 78:8.713087 79:14.335286 80:8.562037 81:5.552081 82:417.0 83:9.291153 84:12.989067 85:8.994183 86:10.251712 87:9.119736 88:13.704722 89:9.083848 90:5.0786834 91:4074.0 92:9.3569145 93:13.082547 94:9.013313 95:10.313046 96:9.410319 97:14.031446 98:9.158228 99:5.3676524 100:3622.0 101:9.758806 102:18.50928 103:5.686489 104:8.085944 105:8.37732 106:15.537158 107:9.518146 108:5.836766 109:87.0 110:0 111:3 112:0 113:0 114:0 115:53 116:2 117:2 118:9693 119:2 120:1.295276E-4 121:2.4846808E7 122:10
0 1:1090311 2:4.9832096 3:4.8724627 4:4.744863 5:5.363034 6:7.268688 7:13.654649 8:4.9566574 9:1.1947924 10:28.0 11:6.598146 12:11.971948 13:1.3817906 14:1.981788 15:4.0055294 16:16.96273 17:6.2089005 18:5.2768097 19:2.0 20:6.790642 21:11.7454815 22:1.7986724 23:2.5756114 24:3.2440546 25:17.218998 26:6.528393 27:7.221268 28:2.0 29:0.0 30:0.0 31:0.0 32:0.0 33:0.0 34:0.0 35:0.0 36:0.0 37:0.0 38:0.0 39:0.0 40:0.0 41:0.0 42:0.0 43:0.0 44:0.0 45:0.0 46:0.0 47:0.0 48:0.0 49:0.0 50:0.0 51:0.0 52:0.0 53:0.0 54:0.0 55:0.0 56:0.0 57:0.0 58:0.0 59:0.0 60:0.0 61:0.0 62:0.0 63:0.0 64:0.0 65:9.468631 66:14.087852 67:9.762392 68:12.888271 69:5.1281805 70:17.67822 71:9.539157 72:6.723405 73:7.0 74:6.0780387 75:10.597749 76:6.5000153 77:8.797136 78:3.46321 79:10.750796 80:6.1382794 81:5.7980456 82:3.0 83:8.999931 84:12.723371 85:11.051114 86:13.2150345 87:5.7876873 88:17.779926 89:9.064838 90:6.311686 91:10.0 92:9.011468 93:12.753544 94:10.907694 95:13.146717 96:5.8060293 97:17.785282 98:9.075429 99:6.319274 100:10.0 101:5.266149 102:11.595957 103:5.033003 104:7.2034454 105:3.4946992 106:11.877771 107:5.3800707 108:6.5100336 109:1.0 110:0 111:3 112:0 113:0 114:0 115:92 116:3 117:1 118:770 119:5 120:4.938E-7 121:1.8903694E7 122:552